requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,189 @@
+
+## Dismissed solution B: Dashboards/insights/etc. sharing between projects
+
+What if instead of copying whole trees of entities we shared them directly? This could be even less work for the user – no need to sync manually – and for us – no need to copy dozens of entities to share a single dashboard between two projects. Also no extra work in terms of permissioning or ingestion.
+
+#### Why not?
+
+The UX would still be a little clunky, with a need to manually set up multiple full projects for each environment. It's the taxonomy that's the killer though – shared taxonomy seems to be a relatively common ask in itself, but it's also key for analysis in an implicit way. Group type indexes in particular MUST match exactly, or insights will show results for completely different data. Property types matter – numbers are treated differently from strings, and then there are also dates. A deeper level of integration is needed for this to work reliably.
+
+## Proposed solution: Environments
+
+We can make environments a first-class entity in PostHog – each project would be subdivided into one or more environments. The *data* would be isolated between projects, but the *taxonomy* and *analysis setup* would be shared (see "Summarizing needs" for what should be isolated and what should be shared exactly).
+
+To support the most common workflow of drafting dashboards/etc. in dev/staging and then "promoting" to production, entities such as dashboards would be environment-specific by default, but could be rolled out to the whole project with two clicks. When rolled out, the entity's dependencies would be made project-wide as well (e.g. for a cohort, all actions and other cohorts used in the definition). This would involve no copying though, so there would be no effort needed to sync things.
+
+To support large customers with access control needs, we'd need to add environment-specific project membership to the existing project access control system.
+
+To support the common feature flags workflow of rollout conditions being different between dev and production, we'd make flag rollout groups always environment-specific.
+
+To ensure straightforward yet complete separation of data at the ingestion level (incl. persons), *as well as* steady querying performance, each environment would have its own `team_id`. Due to this requirement, environments would have to be predefined by users and not created at ingestion time, thus each environment would also have its own SDK key (aka "project API key" currently). That seems like a safe choice in terms of the UX.
```

Comment:
We want to have multiple api keys per project - sources in CDP.

Reply:
btw I wrote up a quick requirements doc draft for this: https://docs.google.com/document/d/150Uly1PpjAX5C_uZH9qkqGltrlBKyiEhbI9WVX9eW9s

Reply:
Makes sense, and we talked about this sync with Tiina. Added a step early on where we split out project API keys from `Team` into an `SDKKey` model

Reply:
>> new table with a foreign key to a single environment.
> to a single team?

to a single data_team_id, so an environment. We'll want dev and prod to have separate capture tokens, to route events to the right environment within a project.

Reply:
Yup